now with just 2 files and no old stuff from month ago

nicedevil007 opened the pull request:
## Description

The ADMIN_TOKEN should made hashed instead of plain-text. Further information can be found here: https://github.com/dani-garcia/vaultwarden/wiki/Enabling-admin-page#secure-the-admin_token

We get this on the `/admin` page as a warning on login.
![image](https://user-images.githubusercontent.com/17103076/230784112-ad06f829-d3c3-4a13-a12b-d58f61e8259d.png)

If we want to fix it with a already running installation, we just have to run the update script to setup a new or the old admin token. Then this will be hashed with the `owasp` preset and inserted on the right places of the config files.

Here a few screenshots during the setup:
![image](https://user-images.githubusercontent.com/17103076/230784304-8bdf8164-8e0b-43b9-90c9-0e3cca779cef.png)

This will be shown if a user doesn't setup an ADMIN_TOKEN during the installation :)
![image](https://user-images.githubusercontent.com/17103076/230784328-a489e50d-05c3-4be7-835c-9650712a16b4.png)

## Type of change

- [X] Bug fix 
- [X] New feature

nicedevil007 committed:
now with just 2 files and no old stuff from month ago

tteck commented on ct/alpine-vaultwarden.sh:
```diff
@@ -88,3 +103,4 @@ description
 msg_ok "Completed Successfully!\n"
 echo -e "${APP} should be reachable by going to the following URL.
          ${BL}http://${IP}:8000${CL} \n"
+03713361530
```
What is this?

nicedevil007 committed:
where did that number come from o_O

got rid of it

tteck commented:
Breaking the visual flow of the script, the whiptail used in the installation process should be avoided. My suggestion is to refrain from setting an admin token during installation and instead utilize the update menu to **set** or reset an admin token.

tteck commented on ct/alpine-vaultwarden.sh:
```diff
@@ -73,7 +73,22 @@ function update_script() {
+      if NEWTOKEN=$(whiptail --passwordbox "Setup your ADMIN_TOKEN (make it strong)" 10 58 3>&1 1>&2 2>&3); then
+        if [[ -z "$NEWTOKEN" ]]; then exit-script; fi      
+        ADMINTOKEN=$(echo -n ${NEWTOKEN} | argon2 "$(openssl rand -base64 32)" -e -id -k 19456 -t 2 -p 1)
+        if [[ -f /var/lib/vaultwarden/config.json ]]; then 
```
Why this? There is no file `/var/lib/vaultwarden/config.json`

nicedevil007 replied:
that one I found out during my testing with those 9 files :)

If you setup vaultwarden with the script the first time then the /etc/conf.d/vaultwarden will be read (I mean it will be read all the time if you NEVER changed anything inside the webui/admin panel).

If you ever changed ONE setting on the admin page, this file will be created and then this one is the mainfile.
Thats why we have to change the ADMIN_TOKEN there if anyone changed his settings :)

Small question before I start the part with "no setup during first installation": The whiptail menu during the update script is ok?

I can do a read command like you did on influxdb script for vaultwarden installation ofc... just thought that whiptail looks best and feels more "complete" when its done during the installation process?

nicedevil007 replied:
> 

Ok new commit where I removed the admin token during first installation.
Added the argon2 check during the update process (like you did with newt package)... so if a user starts this script and is using an "older" version of the alpine installation it will be added now
There was a missing "," at the end of the admin_token line in `config.json`, also fixed

tteck replied:
I like the way you're doing the whiptail in the update portion.
Can `/etc/conf.d/vaultwarden` be eliminated in favor of  `/var/lib/vaultwarden/config.json` during installation?

nicedevil007 replied:
> I like the way you're doing the whiptail in the update portion. Can `/etc/conf.d/vaultwarden` be eliminated in favor of `/var/lib/vaultwarden/config.json` during installation?

I tested it but it seems like it needs at least all the 4 lines except the admintoken line... I will test further if it is mandatory if we get those lines inside the config.json as well but I guess no...

tho the only thing we can get rid of is the admin token from this, don't know if it is worth it to make an other if/then loop to take care of this?

nicedevil007 replied:
have an Idea on making the script look cleaner on the update script... give me 5 more minutes please :)

tteck replied:
I doubt a symbolic link would work either

nicedevil007 replied:
the user can change stuff on the admin-page and the 2 files are not the same from content side.... a symbolic link won't help here (if I understand what you mean with that... ln command right?)

nicedevil007 committed:
removed Admintoken on INSTALL

also fixed the missing "," after the ADMINTOKEN line

added a check if argon2 is installed or not during update

tteck commented on install/alpine-vaultwarden-install.sh:
```diff
@@ -20,15 +20,15 @@ $STD apk add openssl
 
 msg_info "Installing Alpine-Vaultwarden"
 $STD apk add vaultwarden
-cat <<EOF >/etc/conf.d/vaultwarden
```
?? removed

nicedevil007 replied:
fixed

nicedevil007 committed:
1 line too much :D

nicedevil007 commented:
should I add a hint here for the user that let him know how to setup an admin_token if he wants to?

```python
echo -e "${APP} should be reachable by going to the following URL.
         ${BL}http://${IP}:8000${CL} \n"
```

something like this

```python
echo -e "ADMIN_TOKEN can be setup by running the Update-Script inside the LXC.
${APP} should be reachable by going to the following URL.
         ${BL}http://${IP}:8000${CL} \n"
```

tteck commented:
> should I add a hint here for the user that let him know how to setup an admin_token if he wants to?
> 
> ```python
> echo -e "${APP} should be reachable by going to the following URL.
>          ${BL}http://${IP}:8000${CL} \n"
> ```
> 
> something like this
> 
> ```python
> echo -e "ADMIN_TOKEN can be setup by running the Update-Script inside the LXC.
> ${APP} should be reachable by going to the following URL.
>          ${BL}http://${IP}:8000${CL} \n"
> ```

I'd just leave that to the instructions

nicedevil007 committed:
much cleaner

we don't need to overwrite the other config variables inside the vaultwarden file

nicedevil007 commented:
now it is more consistent to the other sed command for the other file :)

tteck commented:
Is this any better?

```
  if NEWTOKEN=$(whiptail --passwordbox "Setup your ADMIN_TOKEN (make it strong)" 10 58 3>&1 1>&2 2>&3); then
    if [[ -z "$NEWTOKEN" ]]; then exit-script; fi
  TOKEN=$(echo -n ${NEWTOKEN} | argon2 "$(openssl rand -base64 32)" -e -id -k 19456 -t 2 -p 1)
    if [[ ! -f /var/lib/vaultwarden/config.json ]]; then 
      echo '{"ADMIN_TOKEN": ""}' > /var/lib/vaultwarden/config.json
    fi
    sed -i "s/ADMIN_TOKEN='.*'/ADMIN_TOKEN='${TOKEN}'/" /var/lib/vaultwarden/config.json
    sed -i "s/export ADMIN_TOKEN='.*'/export ADMIN_TOKEN='${TOKEN}'/" /etc/conf.d/vaultwarden
    rc-service vaultwarden restart -q
  fi
```

nicedevil007 commented:
> Is this any better?
> 
> 
> 
> ```
> 
>   if NEWTOKEN=$(whiptail --passwordbox "Setup your ADMIN_TOKEN (make it strong)" 10 58 3>&1 1>&2 2>&3); then
> 
>     if [[ -z "$NEWTOKEN" ]]; then exit-script; fi
> 
>   TOKEN=$(echo -n ${NEWTOKEN} | argon2 "$(openssl rand -base64 32)" -e -id -k 19456 -t 2 -p 1)
> 
>     if [[ ! -f /var/lib/vaultwarden/config.json ]]; then 
> 
>       echo '{"ADMIN_TOKEN": ""}' > /var/lib/vaultwarden/config.json
> 
>     fi
> 
>     sed -i "s/ADMIN_TOKEN='.*'/ADMIN_TOKEN='${TOKEN}'/" /var/lib/vaultwarden/config.json
> 
>     sed -i "s/export ADMIN_TOKEN='.*'/export ADMIN_TOKEN='${TOKEN}'/" /etc/conf.d/vaultwarden
> 
>     rc-service vaultwarden restart -q
> 
>   fi
> 
> ```

No it is not, if you check if the file doesn't exists you don't have to echo anything in it.

But if the file exists you should keep all the other 20-30 lines in it and don't overwrite them, that's why I used the sed command.

And vise versa we don't need to change anything in the conf.d file if the user already made changes in the webui because then the admin_token will be used from the config.json and the remaining 4 settings from the conf.d file are still needed in that case

tteck commented:
By reading the code, you will notice that it creates the "/var/lib/vaultwarden/config.json" file if it's not already there. This ensures that all installations, whether old or new, are identical. The two sed commands achieve the same result as your four commands but in a more succinct manner.

tteck commented:
This would be my suggestion.

```
      if NEWTOKEN=$(whiptail --passwordbox "Set the ADMIN_TOKEN" 10 58 3>&1 1>&2 2>&3); then
        if [[ -z "$NEWTOKEN" ]]; then exit; fi
        if ! command -v argon2 >/dev/null 2>&1; then apk add argon2 &>/dev/null; fi
        TOKEN=$(echo -n ${NEWTOKEN} | argon2 "$(openssl rand -base64 32)" -t 2 -m 16 -p 4 -l 64 -e)
        if [[ ! -f /var/lib/vaultwarden/config.json ]]; then
          echo '{"ADMIN_TOKEN": ""}' >/var/lib/vaultwarden/config.json
          chmod 777 /var/lib/vaultwarden/config.json
        fi
        sed -i "s|ADMIN_TOKEN=.*|ADMIN_TOKEN='${TOKEN}'|" /var/lib/vaultwarden/config.json
        sed -i "s|export ADMIN_TOKEN=.*|export ADMIN_TOKEN='${TOKEN}'|" /etc/conf.d/vaultwarden
        rc-service vaultwarden restart -q
      fi
```

nicedevil007 commented:
> This would be my suggestion.
> 
> ```
>       if NEWTOKEN=$(whiptail --passwordbox "Set the ADMIN_TOKEN" 10 58 3>&1 1>&2 2>&3); then
>         if [[ -z "$NEWTOKEN" ]]; then exit; fi
>         if ! command -v argon2 >/dev/null 2>&1; then apk add argon2 &>/dev/null; fi
>         TOKEN=$(echo -n ${NEWTOKEN} | argon2 "$(openssl rand -base64 32)" -t 2 -m 16 -p 4 -l 64 -e)
>         if [[ ! -f /var/lib/vaultwarden/config.json ]]; then
>           echo '{"ADMIN_TOKEN": ""}' >/var/lib/vaultwarden/config.json
>           chmod 777 /var/lib/vaultwarden/config.json
>         fi
>         sed -i "s|ADMIN_TOKEN=.*|ADMIN_TOKEN='${TOKEN}'|" /var/lib/vaultwarden/config.json
>         sed -i "s|export ADMIN_TOKEN=.*|export ADMIN_TOKEN='${TOKEN}'|" /etc/conf.d/vaultwarden
>         rc-service vaultwarden restart -q
>       fi
> ```

_before I go in detail about your suggestion, I have a question for you that I want to understand :)
You have a check for the newt package that is in my opinion much easier to read, why you switched to this code here for the argon2 package?_
```python
if ! command -v argon2 >/dev/null 2>&1; then apk add argon2 &>/dev/null; fi
```

_here for comparison:_
```python
if ! apk -e info newt >/dev/null 2>&1; then apk add -q newt; fi
```

_wouldn't it be easier to read and be more consistent with this?_
```python
if ! apk -e info argon2 >/dev/null 2>&1; then apk add -q argon2; fi
```

my 2 SED commands for one file was just because I didn't find the right syntax to get all done in a oneiner, but you got that point. I checked the right permissions for the file and ownership to make it feel vaultwarden it got the file created by itself. Your syntax for the file isn't correct and will lead to a not working vaultwarden.

The `conf.d` file needs `ADMIN_TOKEN=''` and the `config.json` needs `"admin_token": "",`

If we follow your suggestion I would recommend this code here but then we have a small coding issue. If there is no config.json file, then we create a new JSON object that is filled with the value for the token. In that case we don't need the `,` at the end of the json object. If the file exists we will need it unless we put the admin_token into the last line (then we have to remove it first)... makes things a bit more complicated: BUT... that is only a small mistake we see as coders. It will still work I tested it but don't know if that will be the case for the future.

So here my suggestion with your code and right syntax for the file and ownership:

```python
if NEWTOKEN=$(whiptail --passwordbox "Set the ADMIN_TOKEN" 10 58 3>&1 1>&2 2>&3); then
  if [[ -z "$NEWTOKEN" ]]; then exit; fi
  if ! command -v argon2 >/dev/null 2>&1; then apk add argon2 &>/dev/null; fi
  TOKEN=$(echo -n ${NEWTOKEN} | argon2 "$(openssl rand -base64 32)" -t 2 -m 16 -p 4 -l 64 -e)
  if [[ ! -f /var/lib/vaultwarden/config.json ]]; then
    echo '{"admin_token": ""}' >/var/lib/vaultwarden/config.json
    chown 100:101 /var/lib/vaultwarden/config.json
    chmod 644 /var/lib/vaultwarden/config.json
  fi
  sed -i "s|\"admin_token\": .*|\"admin_token\": \"${TOKEN}\",|" /var/lib/vaultwarden/config.json
  sed -i "s|export ADMIN_TOKEN=.*|export ADMIN_TOKEN='${TOKEN}'|" /etc/conf.d/vaultwarden
  rc-service vaultwarden restart -q
fi
```

And here me suggestion with the sed command I used before without the issue from above, much smaller now with your syntax for sed and we don't have to set permissions on the file

```python
if NEWTOKEN=$(whiptail --passwordbox "Setup your ADMIN_TOKEN (make it strong)" 10 58 3>&1 1>&2 2>&3); then
  if [[ -z "$NEWTOKEN" ]]; then exit-script; fi
  if ! command -v argon2 >/dev/null 2>&1; then apk add argon2 &>/dev/null; fi
  TOKEN=$(echo -n ${NEWTOKEN} | argon2 "$(openssl rand -base64 32)" -e -id -k 19456 -t 2 -p 1)
  if [[ ! -f /var/lib/vaultwarden/config.json ]]; then
    sed -i "s|export ADMIN_TOKEN=.*|export ADMIN_TOKEN='${TOKEN}'|" /etc/conf.d/vaultwarden
  else
    sed -i "s|\"admin_token\": .*|\"admin_token\": \"${TOKEN}\",|" /var/lib/vaultwarden/config.json
  fi
  rc-service vaultwarden restart -q
fi
```

In my opinion we shouldn't care about the 2 files with 2 times the admin_token in it. Because that is the normal behaviour of vaultwarden. If a user changes something (like SMTP Server f.e. or the admin_token itself through the webui) the old conf.d file will be left behind (except those 4 needed lines) and everything else will be taken from the config.json.

nicedevil007 committed:
sed syntax

tteck commented:
> before I go in detail about your suggestion, I have a question for you that I want to understand :)
You have a check for the newt package that is in my opinion much easier to read, why you switched to this code here for the argon2 package?

Consistency is important in programming, and both code snippets achieve the same goal functionally. However, the second code snippet is a more versatile solution since it verifies the availability of the argon2 command in the shell's search path, making it useful in situations where the package manager is different or unavailable. You can choose the method that suits you best. Once you make the necessary changes, I can merge the code. Personally, I find coding for Alpine Linux unpleasant due to its minimalist approach.

tteck commented:
I remember now, `newt` is not a command. It's a package.

nicedevil007 commented:
> I remember now, `newt` is not a command. It's a package.

yeah then we had to check whiptail as the command... I guess it is a good idea to stay at your recommendation because of the availability of the package manager... so my last commit already included everything, you can review it :)

https://github.com/tteck/Proxmox/pull/1314/commits/af7b8e66e72884b7d1e0878d9198d72f1f495047
